server/v2/stf/stf_router.go

```diff
@@ -139,39 +140,71 @@ func msgTypeURL(msg gogoproto.Message) string {
 	return gogoproto.MessageName(msg)
 }
 
-var _ router.Service = (*Router)(nil)
+var _ router.Service = (*coreRouterImpl)(nil)
 
-// Router implements the STF router for msg and query handlers.
-type Router struct {
+// coreRouterImpl implements the STF router for msg and query handlers.
+type coreRouterImpl struct {
 	handlers map[string]appmodulev2.Handler
 }
 
-func (r Router) CanInvoke(_ context.Context, typeURL string) error {
+func (r coreRouterImpl) CanInvoke(_ context.Context, typeURL string) error {
+	// trimming prefixes is a backwards compatibility strategy that we use
+	// for baseapp components that did routing through type URL rather
+	// than protobuf message names.
+	typeURL = strings.TrimPrefix(typeURL, "/")
 	_, exists := r.handlers[typeURL]
 	if !exists {
 		return fmt.Errorf("%w: %s", ErrNoHandler, typeURL)
 	}
 	return nil
 }
 
-func (r Router) InvokeTyped(ctx context.Context, req, resp gogoproto.Message) error {
+func (r coreRouterImpl) InvokeTyped(ctx context.Context, req, resp gogoproto.Message) error {
 	handlerResp, err := r.InvokeUntyped(ctx, req)
 	if err != nil {
 		return err
 	}
-	merge(handlerResp, resp)
-	return nil
-}
-
-func merge(src, dst gogoproto.Message) {
-	reflect.Indirect(reflect.ValueOf(dst)).Set(reflect.Indirect(reflect.ValueOf(src)))
+	return merge(handlerResp, resp)
 }
 
-func (r Router) InvokeUntyped(ctx context.Context, req gogoproto.Message) (res gogoproto.Message, err error) {
+func (r coreRouterImpl) InvokeUntyped(ctx context.Context, req gogoproto.Message) (res gogoproto.Message, err error) {
 	typeName := msgTypeURL(req)
 	handler, exists := r.handlers[typeName]
 	if !exists {
 		return nil, fmt.Errorf("%w: %s", ErrNoHandler, typeName)
 	}
 	return handler(ctx, req)
 }
+
+func merge(src, dst gogoproto.Message) error {
+	if src == nil {
+		return fmt.Errorf("source message is nil")
+	}
+	if dst == nil {
+		return fmt.Errorf("destination message is nil")
+	}
+
+	srcVal := reflect.ValueOf(src)
+	dstVal := reflect.ValueOf(dst)
```

Comment:
do we want to avoid reflection here?

Reply:
cannot be avoided, unless we marshal unmarshal